// expect-enzyme/dist/index.js
'use strict';

var _expect = require('expect');

var _expect2 = _interopRequireDefault(_expect);

var _assertions = require('./assertions');

var _assertions2 = _interopRequireDefault(_assertions);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// eslint-disable-next-line
module.exports = function enzymify() {
  var asserted = (0, _expect2.default)();

  return (0, _assertions2.default)({
    // Custom.
    toNotHaveRendered: asserted.toNotHaveRendered,
    toNotHaveContext: asserted.toNotHaveContext,
    toHaveRendered: asserted.toHaveRendered,
    toNotHaveStyle: asserted.toNotHaveStyle,
    toNotHaveClass: asserted.toNotHaveClass,
    toNotHaveProps: asserted.toNotHaveProps,
    toNotHaveState: asserted.toNotHaveState,
    toNotHaveProp: asserted.toNotHaveProp,
    toHaveContext: asserted.toHaveContext,
    toHaveClass: asserted.toHaveClass,
    toHaveState: asserted.toHaveState,
    toHaveProps: asserted.toHaveProps,
    toHaveStyle: asserted.toHaveStyle,
    toHaveProp: asserted.toHaveProp,

    // Built-in.
    toNotContain: asserted.toNotContain,
    toNotExist: asserted.toNotExist,
    toContain: asserted.toContain,
    toNotBeAn: asserted.toNotBeAn,
    toNotBeA: asserted.toNotBeA,
    toExist: asserted.toExist,
    toBeAn: asserted.toBeAn,
    toBeA: asserted.toBeA
  });
};
